utils/polus-filepattern-util/filepattern/classes.py

```diff
@@ -0,0 +1,269 @@
+
+        Args:
+            files: A list of file dictionaries
+
+        Returns:
+            An output file name
+        """
+        
+        if len(files) == 0:
+            files = self.files
+            files = get_matching(files,self.var_order,**{k.upper():v for k,v in self.uniques.items()})
+        # else:
+        #     files,uniques = _parse(parse_filename,
+        #                            [file['file'] for file in files],
+        #                            self.pattern,
+        #                            self.regex,
+        #                            self.variables,
+        #                            self.var_order)
+        
+        vals = {v:set() for v in self.var_order}
+        for file in files:
+            for k,v in file.items():
+                if k not in self.var_order:
+                    continue
+                vals[k].add(v)
+        
+        kwargs = {}
+        for k,v in vals.items():
+            v = list(v)
+            if len(v) == 1 and v[0] != -1:
+                kwargs[k] = v[0]
+                
+        return output_name(self.pattern,files,kwargs)
+
+    # Get filenames matching values for specified variables
+    def get_matching(self,**kwargs):
+        """ Get all filenames matching specific values
+    
+        This function runs the get_matching function using the objects file
+        dictionary. For more information, see :any:`get_matching`.
+
+        Args:
+            **kwargs: One of :any:`VARIABLES`, must be uppercase, can be single
+                values or a list of values
+                
+        Returns:
+            A list of all files matching the input values
+        """
+        # get matching files
+        files = get_matching(self.files,self.var_order,out_var=None,**kwargs)
+        return files
+
+    def __iter__(self):
+        
+        group_by = self._group_by
+        kwargs = self._kwargs
+        self._group_by = None
+        self._kwargs = None
+        
+        if kwargs == None:
+            raise SyntaxError('Cannot directly iterate over a FilePattern object. Call it (i.e. for i in fp())')
+
+        # If self.files is a list, no parsing took place so just loop through the files
+        if isinstance(self.files,list):
+            for f in self.files:
+                yield [f]
+            return
+
+        # Generate the values to iterate through
+        iter_vars = {}
+        for v in self.var_order:
+            if v in group_by:
+                continue
+            elif v.upper() in kwargs.keys():
+                if isinstance(kwargs[v.upper()],list):
+                    iter_vars[v] = copy.deepcopy(kwargs[v.upper()])
+                else:
+                    iter_vars[v] = [kwargs[v.upper()]]
+            else:
+                iter_vars[v] = copy.deepcopy(self.uniques[v])
+        
+        # Find the shallowest variable in the dictionary structure
+        shallowest = None
```

Comment:
Shouldn't a dict always have a shallowest variable? Is `shallowest` the best name for this variable in this case? or perhaps the comment should be rephrased to indicate what's actually going on.

Reply:
Actually, this may be one of those instances where my thinking is backwards relative to everyone else. When I was writing this, it made sense to call the `shallowest` variable the one closest to the list of file dictionaries. In reality, I think it might make sense to call it `deepest`. I added comments for clarity.